### ISSUE TITLE
Built-in feature for showing a session timeout warning

### ISSUE THREAD

**Issue by rolfsmeds** [vaadin/flow#15577]
### Describe your motivation

WCAG 2.1 accessibility guidelines (level A) stipulate that the user should be warned (at least 20s) before a time limit, such as a session timeout, is reached, and given an opportunity to extend the time limit (https://www.w3.org/TR/WCAG21/#timing-adjustable)

Although it's possible to implement this from scratch, it would be beneficial for a session-based framework like Vaadin to provide a simple default mechanism for this, complete with some kind of UI for showing the warning and extending the time limit.

### Describe the solution you'd like

Some kind of mechanism for automatically showing a warning (rendered as an overlay on top of the UI) and providing the user a way to reset the timeout counter, thereby extending the session time limit.

- opt-in (or opt-out) to avoid (or revert) breaking changes caused by this
- i18n API for configuring the message
- the default UI should probably be an overlay rendered on top of the app UI
- a11y needs to be considered here (focus should probably not move automatically to the warning, as that would be disruptive, but keyboard and screen reader users need a quick and easy way to trigger the time extension button)
- a way to replace the default UI with something custom, at least a custom overlay but ideally a way to instead render it inline in the UI layout if so desired

**Comment by knoobie** [comment]
> a11y needs to be considered here (focus should probably not move automatically to the warning, as that would be disruptive, but keyboard and screen reader users need a quick and easy way to trigger the time extension button)

Normally the session timeout should only occur if the user did NOT interact with the application for a longer period of time, so changing focus sounds right to me. I would even argue to use role alert dialog to make sure it's really eager published to the user. Even tho I don't believe people going back to just press the button 20 seconds before auto-logout comes into play. 

We normally have 30 or 60min visual timer (triggered/refreshed by push when interacting with the app) on our apps, once it reaches 5min a dialog is shown to inform the user. If no action was triggered by the user we log them out at 1min and out-save the current open form.

**Comment by rolfsmeds** [comment]
Good point @knoobie – active user means no timeout, thus no disruption issue.
And that does mean that we could even use the native alert.
The point in time at which the warning is shown should of course be customizable, since the optimum can vary a lot between applications (and the default should definitely be more than the WCAG 20s minimum).

The visual timer would of course need to be a custom UI feature, but certainly the framework could provide a feature for making that as easy to implement as possible.

**Comment by MarcinVaadin** [comment]
Isn't https://vaadin.com/directory/component/idle-notification/overview enough?

**Comment by rolfsmeds** [comment]
I'm not familiar with the details of that add-on, but it could be something very much like that, except part of the framework rather than an add-on.

**Comment by cyberoblivion** [comment]
This is something we've implemented internally. But, would love to see this as a built in feature!

**Comment by mstahv** [comment]
The IdleNotification add-on seems to be also relying on the "hearbeat" functionality, which is not exactly designed for thiss and can hopefully be removed from the product at some point. Looked a bit into the "Vaadin" namespace using browser inspector and found one way to get notified about client-server activity. This allowed me to build a system that works on the client-side, works across multiple browser tabs and contains a special "execute this before killing the session" action that can be optionally used.

@knoobie, dare to try this at some point? Maybe we can lure Rolf to pick this task if we have a good enough prototype ready 🤓

https://github.com/viritin/flow-viritin/compare/v24...feature/session-timeout-warning-dialog

<img width="800" alt="Screenshot 2023-06-01 at 15 49 56" src="https://github.com/vaadin/flow/assets/1478052/63d6672a-61f6-49de-9009-6dd11b50fb70">

**Comment by mstahv** [comment]
Here is one alternative approach, built completely on the server side. This requires push to be on: https://gist.github.com/mstahv/956b4682fd3b16ffb5d1a0638892893a
